### PR TITLE
Sequencing events

### LINKED ISSUE #88: Sequencing support (a.k.a. sequence events)

### PR DESCRIPTION
**Pull request by mitschabaude**
- Closes #88 
- RFC: https://github.com/o1-labs/snarkyjs/issues/265

### CODE REVIEW

File: src/lib/zkapp.ts [hunk 1]
```diff
@@ -370,6 +450,134 @@ export class SmartContract {
...
+      { maxTransactionsWithUpdates = 32 } = {}
+    ): { state: S; stateHash: Field } {
+      if (updateLists.length > maxTransactionsWithUpdates) {
+        throw Error(
+          `stateUpdate.applyUpdates: Exceeded the maximum number of lists of updates, ${maxTransactionsWithUpdates}.
+Use the optional \`maxTransactionsWithUpdates\` argument to increase this number.`
+        );
+      }
+      let methodData = contract.analyzeMethods();
+      let possibleUpdatesPerTransaction = [
```
Comment (mitschabaude) [inline src/lib/zkapp.ts:524]: one of the magical things this function does is it knows how many events each of the smart contract methods emits, so it can perform computations for each of those numbers-of-events and then pick the actual one.
to enable that, I added `SmartContract.analyzeMethods` in this file and `synthesizeMethodArguments` in proof_systems.ts

File: src/examples/state_update_rollup.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      state.counter = state.counter.add(update);
+      return state;
+    },
+  });
+
+  // on-chain version of our state. it will typically lag behind the
+  // version that's implicitly represented by the list of updates
+  @state(Field) counter = State<Field>();
+  // helper field to store the point in the update history that our on-chain state is at
+  @state(Field) stateHash = State<Field>();
```
Comment (bkase) [inline src/examples/state_update_rollup.ts:37]: let's change this to whatever you pick on the RFC-side like `actionsHash` or `actionsCommitment`

File: src/examples/state_update_rollup.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+  // emits an update
+  @method incrementCounter(increment: Field) {
+    this.stateUpdate.emit(increment);
+  }
+
+  @method rollupStateUpdate() {
```
Comment (bkase) [inline src/examples/state_update_rollup.ts:44]: -> reduce

File: src/lib/circuit_value.test.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { Bool, Circuit, isReady, shutdown, Int64 } from '../../dist/server';
+
+describe('circuit', () => {
```
Comment (bkase) [inline src/lib/circuit_value.test.ts:3]: can you add tests expecting failures if there are more than one true and zero true?

**Review by bkase — COMMENTED** [review]
This still needs updating to the latest version of the RFC before we should land it even under the "experimental" flag -- and don't forget to change the names of things everywhere! I stopped mentioning it after I realized you haven't updated this yet.

File: src/lib/circuit_value.test.ts (hunk 4 above)
Comment (mitschabaude) [reply src/lib/circuit_value.test.ts:3]: done. for some reason returning 0 for zero true seemed natural to me. but I can also throw in that case if you think that's better

File: src/lib/circuit_value.ts [hunk 5]
```diff
@@ -467,3 +469,39 @@ function circuitValueEquals<T>(a: T, b: T): boolean {
     ([key, value]) => key in b && circuitValueEquals((b as any)[key], value)
   );
 }
+
+function pickOne<T, A extends AsFieldElements<T>>(
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/lib/circuit_value.ts]: switch is great

File: src/lib/circuit_value.test.ts (hunk 4 above)
Comment (bkase) [reply src/lib/circuit_value.test.ts:3]: good point yeah 0 is fine, but can you add a test for the zero case succeeding too?

File: src/lib/zkapp.ts [hunk 6]
```diff
@@ -370,6 +441,119 @@ export class SmartContract {
...
+type Reducer<Action> = { actionType: AsFieldElements<Action> };
+
+type ReducerReturn<Action> = {
+  dispatch(action: Action): void;
+  reduce<State>(
+    actions: Action[][],
+    stateType: AsFieldElements<State>,
+    reduce: (state: State, action: Action) => State,
+    initial: { state: State; actionsHash: Field },
+    advancedOptions?: { maxTransactionsWithActions?: number }
```
Comment (bkase) [inline src/lib/zkapp.ts:453]: this is nice

Comment (bkase) [reply src/lib/zkapp.ts:453]: probably just `options?` is fine

**Review by bkase — APPROVED** [review]

File: src/examples/state_update_rollup.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+console.log('action 1');
+tx = await Mina.transaction(feePayer, () => {
+  zkapp.incrementCounter();
+  if (!doProofs) zkapp.sign(zkappKey);
+});
+if (doProofs) await tx.prove();
+tx.send();
+// update internal state
+pendingActions.push([INCREMENT]);
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply src/examples/state_update_rollup.ts:109]: yeah I agree. I think this would be achieved if `reducer.getActions` -- which will fetch actions in the real network interaction -- gets them from memory in the LocalBlockchain case
